### ISSUE TITLE
Type annotation error for generic list that includes union

### ISSUE THREAD

**Issue by vishwa-raj** [microsoft/pylance-release#1526]
<h3> Environment data </h3>

-   Language Server version: 2021.6.3
-   OS and version: win32 x64
-   Python version (and distribution if applicable, e.g. Anaconda): 
-   python.analysis.indexing: undefined
-   python.analysis.typeCheckingMode: basic

<h3> Code Snippet / Additional information </h3>

```python
from __future__ import annotations

from typing import TypeVar


NUM_T = TypeVar('NUM_T', int, float)


def funcA(value: list[NUM_T | list[NUM_T]]):
    ...


def funcB(value: list[NUM_T | list[NUM_T]]):
    funcA(value)

```

<h3> Expected behaviour </h3>

No error.

<h3> Actual behaviour </h3>

Error shown on line 14:
```
Argument of type "list[NUM_T@func_B | list[NUM_T@func_B]]" cannot be assigned to parameter "value" of type "list[NUM_T@func_A | list[NUM_T@func_A]]" in function "func_A"
  TypeVar "_T@list" is invariant
    Type "NUM_T@func_B | list[NUM_T@func_B]" cannot be assigned to type "NUM_T@func_A | list[NUM_T@func_A]"
      Type "int* | float*" cannot be assigned to type "NUM_T@func_A | list[NUM_T@func_A]"
        Type "int*" cannot be assigned to type "NUM_T@func_A | list[NUM_T@func_A]"
          "int*" is incompatible with "list[NUM_T@func_A]"
        Type "float*" cannot be assigned to type "NUM_T@func_A | list[NUM_T@func_A]"
          "float*" is incompatible with "list[NUM_T@func_A]"
      Type "list[NUM_T@func_B]" cannot be assigned to type "NUM_T@func_A | list[NUM_T@func_A]"
```

![image](https://user-images.githubusercontent.com/22815753/124455191-e7e84280-dda6-11eb-9600-d7271a81aef0.png)

**Comment by erictraut** [comment]
Why are you using a type variable in this case? Do you mean to use a union type?
```python
NUM_T = Union[int, float]
```

or
```python
T = TypeVar("T")
LIST_T = list[T | list[T]]
NUM_T = Union[LIST_T[int], LIST_T[float]]
```

**Comment by vishwa-raj** [comment]
@erictraut It's a fabricated example from a larger piece of code whose type annotation may look like `list[NUM_T] | list[tuple[NUM_T, NUM_T]]`.
What we need here is to be able to allow both `int` and `float` type inputs.
If I just use `float` or `float | int` instead of `NUM_T`, I cannot pass an input like [1, 2, 3] without raising a type-annotation error as the input is of of type `list[int]` and `list` is invariant.
I can enumerate all combinations of `int` and `float`, like `list[int] | list[float] | list[tuple[int, int]] | list[tuple[int, float]] ...` but that will get unwieldy as the original type annotation include many more combinations.

**Comment by erictraut** [comment]
This will be fixed in the next release. I've improved the heuristics for type matching when the source and destination types are both unions and invariance is enforced. The rules for matching in this case are not well defined and can be ambiguous in some cases where generics are involved in the destination type, but I've improved the heuristics to work in more cases including the one above.

**Comment by bschnurr** [comment]
This issue has been fixed in version 2021.7.3, which we've just released. You can find the changelog here: https://github.com/microsoft/pylance-release/blob/main/CHANGELOG.md#202173-14-july-2021
